File: Advance-Data-Structure/linked_list.cpp
#include <bits/stdc++.h>
using namespace std;

struct Node {
    int data;
    Node* next;
};

void InsertBegin(Node** head, int data) {
    Node* temp = new Node();
    temp->data = data;
    temp->next = NULL;
    if (*head != NULL) temp->next = *head;
    *head = temp;
}

void Insert(Node** head, int data, int pos) {
    Node* temp1 = new Node();
    temp1->data = data;
    temp1->next = NULL;

    // pos at the beginning
    if (pos == 1) {
        temp1->next = *head;
        *head = temp1;
        return;
    }

    // n'th position
    Node* temp2 = *head;
    for (int i = 1; i < pos-1; i++) temp2 = temp2->next;

    temp1->next = temp2->next;
    temp2->next = temp1;
}

Node* Reverse(Node* head) {
    Node *current, *prev, *next;
    current = head;
    prev = next = NULL;

    while (current != NULL) {
        next = current->next;
        current->next = prev;
        prev = current;
        current = next;
    }

    head = prev;
    return head;
}

Node* ReverseRecursive(Node* head) {
    if (head->next == NULL) {
        return head;
    }
    Node* temp = ReverseRecursive(head->next);
    head->next->next = head;
    head->next = NULL;
    return temp;
}

void Delete(Node** head, int pos) {
    Node* temp1 = *head;

    // delete first node
    if (pos == 1) {
        *head = temp1->next;
        free(temp1); // de-allocate the memory
        return;
    }

    // n'th position
    for (int i = 1; i < pos-1; i++) temp1 = temp1->next;

    Node* temp2 = temp1->next;
    temp1->next = temp2->next;
    free(temp2);
}

void print(Node* temp) {
    while (temp == NULL) return;
    cout << temp->data << " ";
    print(temp->next);
}

void print_recursive(Node* temp) {
    while (temp == NULL) return;
    print_recursive(temp->next);
    cout << temp->data << " ";
}

void printList(Node* head) {
    cout << "\n=> ";
    print(head);
    cout << "\n";
    cout << "<= ";
    print_recursive(head);
    cout << "\n\n";
}

Node* middleNode(Node* head) {
    if (head == NULL) {
        return head;
    }
    
    Node* a = head;
    Node* b = head->next;

    while (b != NULL && b->next != NULL) {
        a = a->next;
        b = b->next->next;
    }

    return a;
}

Node* Merge(Node* a, Node* b) {
    Node* temp = new Node();
    Node* resultList = temp;

    while (a != NULL && b != NULL) {
        if (a->data > b->data) {
            temp->next = a;
            a = a->next;
        }
        else {
            temp->next = b;
            b = b->next;
        }
        temp = temp->next;
    }

    temp->next = (a == NULL) ? b : a;
    return resultList->next;
}

Node* MergeSort(Node* head) {
    if (head == NULL || head->next == NULL) {
        return head;
    }

    Node* middle = middleNode(head);
    Node* secondHalf = middle->next;
    middle->next = NULL; // end of first half

    return Merge(MergeSort(head), MergeSort(secondHalf));
}

int main() {
    Node* head = NULL;
    
    InsertBegin(&head, 1);
    InsertBegin(&head, 5);
    InsertBegin(&head, 10);
    InsertBegin(&head, 3);
    InsertBegin(&head, 21);
    printList(head);

    Insert(&head, 12, 1);
    Insert(&head, 13, 2);
    printList(head);

    Delete(&head, 2);
    Delete(&head, 1);
    printList(head);

    cout << "Sort:\n";
    Node* sorted = MergeSort(head);
    printList(sorted);

    cout << "Reverse:\n";
    Node* rec = ReverseRecursive(head);
    printList(rec);


}

/*

=> 2 3 4 5 6 
<= 6 5 4 3 2 


=> 12 13 2 3 4 5 6 
<= 6 5 4 3 2 13 12 


=> 2 3 4 5 6 
<= 6 5 4 3 2 


=> 6 5 4 3 2 
<= 2 3 4 5 6 

*/

File: Advance-Data-Structure/PYTHON/linked_list.py
class Node:
    def __init__(self, data=None):
        self.data = data
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None

    def insertAtBeginning(self, data):
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node

    def insertAtEnd(self, data):
        new_node = Node(data)

        if not self.head:
            self.head = new_node
            return

        curr = self.head
        while curr.next:
            curr = curr.next

        curr.next = new_node

    def search(self, data):
        if not self.head:
            print('List is empty!')

        curr = self.head
        while curr:
            if curr.data == data:
                return True
            curr = curr.next
        return False

    def deleteNode(self, position):
        if not self.head:
            return

        curr = self.head

        if position == 0:
            self.head = curr.next
            curr = None
            return

        for i in range(position-1):
            curr = curr.next
            if not curr:
                return

        if not curr.next:
            return

        next_node = curr.next.next
        curr.next = None
        curr.next = next_node

    def printList(self):
        if not self.head:
            print('List is empty!')
            return

        curr = self.head
        while curr:
            print(curr.data, end=' --> ')
            curr = curr.next
        print()


if __name__ == '__main__':
    llist = LinkedList()
    print(llist.search(1))
    llist.printList()
    llist.insertAtBeginning(1)
    print(llist.search(1))
    print(llist.search(2))
    llist.printList()
    llist.insertAtBeginning(3)
    llist.insertAtBeginning(7)
    llist.printList()
    llist.insertAtEnd(-1)
    llist.insertAtEnd(-17)
    llist.printList()
    llist.deleteNode(0)
    llist.printList()
    llist.deleteNode(7)
    llist.printList()
    llist.deleteNode(2)
    llist.printList()
    llist.deleteNode(2)
    llist.printList()
